drivers: mbox: add initial drivers for nRF VEVIF & BELLBOARD

gmarull opened the pull request:
Tested with:

```
west build -b nrf54h20pdk/nrf54h20/cpuapp samples/drivers/mbox -T sample.drivers.mbox.nrf54h20_vevif -p
west build -b nrf54h20pdk/nrf54h20/cpuapp samples/drivers/mbox -T sample.drivers.mbox.nrf54h20_bellboard -p
```

Pending:
- DTize mbox API, so that things are more ergonomic (dt_spec from #mbox-cells) see https://github.com/zephyrproject-rtos/zephyr/pull/69390
- Refactor sample to use dt-spec stuff (depends on https://github.com/zephyrproject-rtos/zephyr/pull/69390)
- Test VEVIF+BELLBOARD together (ie bidirectional) (depends on https://github.com/zephyrproject-rtos/zephyr/pull/69390, as current sample doesn't deal with 2 different mboxes now)

gmarull committed:
dts: bindings: mbox: add nordic,nrf-vevif-local|remote

Add a new binding for Nordic VEVIF (VPR Event Interface) (local and
remote modes).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20_enga: add PPR VEVIF nodes

Add a new nodes for PPR's VEVIF. In app cores, VEVIF registers are part
of the VPR peripheral, so it is exposed as a child node (since it
requires its own properties, eg #mbox-cells). In VPR, it's a CPU child
since it's not a memory-mapped peripheral, but used with CSRs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: mbox: add initial driver for nRF VEVIF

Add a mailbox driver for VEVIF. The driver can be built in either
'local' or 'remote' configuration. This depends on the existence of the
'interrupts' property, which signals that the instance is managed
locally. VEVIF is, as expected, always managed by a VPR core.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: misc: nordic_vpr_launcher: initialize earlier

Set driver to initialize at early POST_KERNEL, so that we make sure
other future dependencies priorities (eg VEVIF) do not conflict.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: drivers: mbox: add support for nRF VEVIF

Add support for nRF VEVIF.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: drivers: mbox: allow to configure RX/TX channels

By adding Kconfig options for both samples.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: drivers: mbox: allow to enable RX/TX

This can be useful to provide multiple setups, eg unidirectional
ping-pong.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: drivers: mbox: misc cleanups/enhancements

- Include cleanup
- s/NET/REMOTE (remote core may have nothing to do with NET)
- Improved error reporting

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: drivers: mbox: add support for nRF BELLBOARD

Add support for testing BELLBOARD between cpuapp (local) and cpuppr
(remote).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: mbox: add nordic,nrf-bellboard-local|remote

Add new bindings to describe the BELLBOARD peripheral in its two
_programming modes_ local or remote.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20_enga: add BELLBOARD nodes

Add nodes for APP/RAD BELLBOARD peripherals.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: mbox: nrf_bellboard: add initial driver

Add a new driver for nRF BELLBOARD peripheral, in both local and remote
modes.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

57300 commented on samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay:
```diff
@@ -0,0 +1,10 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+mbox: &cpuapp_bellboard {
+	interrupts = <96 NRF_DEFAULT_IRQ_PRIORITY>;
+	interrupt-names = "bellboard0";
+	status = "okay";
+};
```
These might be worth moving to the default board configuration, sans label.

gmarull replied:
Could you suggest a default IRQ configuration?

57300 replied:
Let's come back to it later.

57300 commented on samples/drivers/mbox/sample.yaml:
```diff
@@ -23,6 +23,47 @@ tests:
+      - nrf54h20pdk_nrf54h20_cpuapp
+    extra_args:
+      mbox_SNIPPET=nordic-ppr
+      mbox_OVERLAY_CONFIG=boards/nrf54h20pdk_nrf54h20_cpuapp_vevif.conf
```
`OVERLAY_CONFIG` is deprecated.
```suggestion
      mbox_EXTRA_CONF_FILE=boards/nrf54h20pdk_nrf54h20_cpuapp_vevif.conf
```

gmarull replied:
done

57300 commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,194 @@
+
+#define BELLBOARD_IRQ_CONNECT(grp, grp_idx)                                                        \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, grp),                                                  \
+		    (static const uint8_t idx_##grp = grp_idx;                                     \
+		     IRQ_CONNECT(DT_INST_IRQ_BY_NAME(0, grp, irq),                                 \
+				 DT_INST_IRQ_BY_NAME(0, grp, priority), bellboard_local_isr,       \
+				 &idx_##grp, 0);),                                                 \
```
Why does the second to last param need to be a pointer to a byte in ROM?

gmarull replied:
indeed, just passed number as address.

57300 reviewed (commented):
LGTM, just minor comments below.

gmarull commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,194 @@
+
+#include <haly/nrfy_bellboard.h>
+
+#define BELLBOARD_GROUPS 4U
```
I couldn't find this definition, is it in a new HAL version? Anyway, used such name for the local define now, so it can be updated later easily.

57300 committed:
boards: nrf54h20pdk_nrf54h20: Update memory map

Add IPC shared memory regions in the global RAM, as well as an enlarged
SRAM region for Application core.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

gmarull commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,217 @@
+
+static uint32_t bellboard_local_max_channels_get(const struct device *dev)
+{
+	uint32_t max_channels = 0U;
+
+	ARG_UNUSED(dev);
+
+	for (uint8_t i = 0U; i < ARRAY_SIZE(grps); i++) {
+		if (grps[i] != NULL) {
+			max_channels += NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT;
+		}
+	}
+
+	return max_channels;
```
indeed, this is wrong, will fix it

anangl commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,217 @@
+
+#define BELLBOARD_GRP_GET_OR_NULL(grp) COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, grp), (&g_##grp), (NULL))
+
+BELLBOARD_GRP_MAYBE_DEFINE(bellboard0)
+BELLBOARD_GRP_MAYBE_DEFINE(bellboard1)
```
This is a bit confusing as `bellboardX` looks like an instance of the BELLBOARD peripheral. I'd suggest to use `irqX` instead.

anangl commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,217 @@
+#define BELLBOARD_IRQ_COUNT 4U
+
+/* BELLBOARD runtime resources for each group */
+struct mbox_bellboard_local_group {
+	mbox_callback_t cb[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
+	void *user_data[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
```
There is no need to use separate arrays of callbacks and user data pointers for particular groups. Just one instance of each is enough.

gmarull replied:
fixed, good catch

anangl commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,217 @@
+
+	ARG_UNUSED(dev);
+
+	ret = bellboard_local_grp_get(id, &grp_idx);
```
This finds only the first group to which a given event is assigned but it is technically possible that one event is assigned to several groups. It would be good to handle this possibility.

gmarull replied:
reworked driver to support multi-irq mapping

anangl commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,217 @@
+
+#include <haly/nrfy_bellboard.h>
+
+#define BELLBOARD_IRQ_COUNT 4U
```
This number can potentially differ between cores in future devices. Perhaps it should be defined in devicetree.

gmarull replied:
that's true, I guess there are 2 options here:

- Add `nordic,interrupt-lines = <N>;` in DT
- Always force `interrupts` to define the N lines, and add `nordic,interrupt-mask` which would indicate enabled lins (eg 0x3 = irq0 and irq1), this would require reworking the driver a bit, and would change how bellboard is used in DT, so it's a good time to decide.

The (2) looks more user-friendly at a first glance, but, AFAIK, the idea of using multiple IRQs is to play with priorities, in which case... one needs to re-define the `interrupts` property entirely. So (1) seems a better approach.

gmarull commented on drivers/mbox/Kconfig.nrf_bellboard:
```diff
@@ -0,0 +1,17 @@
+	bool "nRF BELLBOARD local driver"
+	depends on DT_HAS_NORDIC_NRF_BELLBOARD_LOCAL_ENABLED
+	default y
+
```
fixed

anangl commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,178 @@
+	BELLBOARD_IRQ_GET_OR_ZERO(irq3),
+};
+
+/* BELLBOARD IRQ mappings */
+#define BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(idx)                                                     \
+	COND_CODE_1(DT_INST_PROP_HAS_IDX(0, nordic_interrupt_mapping, idx),                        \
+		    (DT_INST_PROP_BY_IDX(0, nordic_interrupt_mapping, idx)), (0U))
+
+static const uint32_t irq_mappings[BELLBOARD_NUM_IRQS] = {
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(0),
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(2),
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(4),
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(6),
```
I think it's not going to work as intended. For instance, if you use the example from binding:
```
      interrupts = <98 NRF_DEFAULT_IRQ_PRIORITY>,
                   <99 NRF_DEFAULT_IRQ_PRIORITY>;
      interrupt-names = "irq2", "irq3";
      nordic,interrupt-mapping = <0x0000000f 2>, <0x000000f0 3>;
```
you'll get: `irqs = { 0, 0, 98, 99 };` and `irq_mappings = { 0x0000000f, 0x000000f0, 0, 0 };`

anangl replied:
How about using designated initializers and `LISTIFY` to fill `irq_mappings`?

gmarull replied:
good catch, please check current revision

anangl replied:
> good catch, please check current revision

It still won't handle properly the above example. I think something like this should be used:
```
#define EVT_MAPPING_ITEM(idx) DT_INST_PROP_BY_IDX(0, nordic_interrupt_mapping, idx)
#define BELLBOARD_GET_EVT_MAPPING(idx, _)				\
	COND_CODE_1(DT_INST_PROP_HAS_IDX(0, nordic_interrupt_mapping,	\
					 UTIL_INC(UTIL_X2(idx))),	\
		([EVT_MAPPING_ITEM(UTIL_INC(UTIL_X2(idx)))] =		\
			EVT_MAPPING_ITEM(UTIL_X2(idx)),),		\
		())

static const uint32_t evt_mappings[BELLBOARD_NUM_IRQS] = {
	LISTIFY(DT_NUM_IRQS(DT_DRV_INST(0)), BELLBOARD_GET_EVT_MAPPING, ())
};
```
Then it won't be also needed to require that "# interrupt mappings == # interrupts";

gmarull replied:
you're right, updated the driver to user you sample, thanks

gmarull commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,178 @@
+#define BELLBOARD_IRQ_GET_OR_ZERO(irq_)                                                            \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, irq_), (DT_INST_IRQ_BY_NAME(0, irq_, irq)), (0U))
+
+static const uint8_t irqs[BELLBOARD_NUM_IRQS] = {
```
no longer needed in current revision

gmarull commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,178 @@
+
+static void bellboard_local_isr(const void *parameter)
+{
+	uint8_t irq_idx = (uintptr_t)parameter;
```
fixed

gmarull commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,178 @@
+
+			if (*enabled_mask == 0U) {
+				irq_disable(irqs[i]);
+			}
```
irqs are now always enabled, and int_enable/disable is only used

anangl commented on dts/bindings/mbox/nordic,nrf-bellboard-local.yaml:
```diff
@@ -0,0 +1,43 @@
+      interrupt identifier. The bitmask is used to indicate which of the 32
+      possible events are mapped to the given interrupt. For example, given
+      <0x0000000f 2>, the first four events are mapped to interrupt 2
+      (irq2). There needs to be at least one mapping for each interrupt.
```
This last sentence can be dropped now.

anangl commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,170 @@
+BUILD_ASSERT(DT_NUM_IRQS(DT_DRV_INST(0)) <= BELLBOARD_NUM_IRQS, "# interrupt exceeds maximum");
+
+BUILD_ASSERT(DT_INST_PROP_LEN(0, nordic_interrupt_mapping) == 2 * DT_NUM_IRQS(DT_DRV_INST(0)),
+	     "# interrupt mappings != # interrupts");
```
This is also not needed. Just checking if the length is a multiple of 2 should be sufficient.

57300 committed:
boards: nrf54h20pdk: Add default IPC configuration

Add `zephyr,ipc-icmsg*` nodes, with mboxes specified in a common file.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Gerard Marull-Partetas <gerard@teslabs.com>

hubertmis commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:
```diff
@@ -93,6 +96,32 @@
+	nordic,interrupt-mapping = <0x00042001 0>;
+};
+
+&cpusec_cpuapp_ipc {
```
I understand the reason to avoid repeating code and reusing parts of this node between both peers in communication, but the result is a hard-to-understand content of this node because it is scattered across multiple files. It's hard to follow which mbox instance is tx and which is rx.

gmarull replied:
@57300

57300 replied:
I was betting on the mbox placement being easy to deduce from the IPC node label. If it's `cpuxxx_cpuyyy_ipc`, then the first one is `cpuxxx_bellboard/vevif` and the second is `cpuyyy_bellboard/vevif`. If you're building for `cpuxxx`, then the first one is RX and the second is TX (to `cpuyyy`), and vice versa for the other core.

In addition, it should also be possible to tuck away the shared memory configuration in the common file, if we deprecate the `tx/rx-region` properties in favor of `memory-regions` and `memory-region-names`.

This information is not scattered just for the sake of avoiding repetition. It makes sense to track channel and memory configuration at system level, as opposed to individual core level.

But if that's too confusing, then I don't mind going back to duplicating nodes.

hubertmis commented on drivers/mbox/mbox_nrf_vevif_local.c:
```diff
@@ -0,0 +1,123 @@
+		}
+
+		cbs.enabled_mask &= ~BIT(id);
+		irq_disable(vevif_irqs[id]);
```
I would expect a condition here to check if the event being disabled is the last event assigned to this IRQ, before the IRQ gets disabled.
Without this condition, there is a risk that calling this function for event 13 while events 12, 13, and 18 are assigned to IRQ 0, the event 13 gets disabled and a side-effect it disables IRQ line for still active events 12 and 18

masz-nordic replied:
I think this is unnecessary at this point. VEVIF has separate IRQ for every event.

hubertmis replied:
Nice! I thought they share IRQs like in the BELLBOARD

hubertmis commented on samples/drivers/mbox/src/main.c:
```diff
@@ -19,24 +18,25 @@ static void callback(const struct device *dev, uint32_t channel,
+	int ret;
+	const struct device *const dev = DEVICE_DT_GET(DT_NODELABEL(mbox));
 
 	printk("Hello from APP\n");
```
You could consider renaming this to something like MAIN or HOST. This CPU does not have to do anything with an application.

gmarull replied:
Sample needs a bit of love, I’ll rework it as part of the https://github.com/zephyrproject-rtos/zephyr/pull/69390 when I rebase it

hubertmis commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,169 @@
+
+	for (uint8_t i = 0U; i < NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT; i++) {
+		if ((evt_all_mappings & BIT(i)) != 0U) {
+			nrf_bellboard_event_clear(bellboard, nrf_bellboard_triggered_event_get(i));
```
I don't think the driver should clear all the events on initialization.
We can imagine that a CPU X is in a SoftOff state and it is woken up by a BELLBOARD event by another CPU Y. The event is pending. Then the CPU X runs a complete cold-boot procedure, including the bellboard driver initialization. After the initialization is complete and the application code enables bellboard events and IRQs, the pending event must be handled. If the event gets cleared during the driver initialization, the event will be lost and CPU X won't have any idea why it was woken up.

gmarull replied:
Will drop this. Note that I inherited this code from nrfy, so it has always been done like that. I guess it wasn’t the right thing to do.

gmarull replied:
Note: I'll open a follow-up PR resolving any remaining issues after we have a call to discuss more details, since the logic in this PR has been largely based on what we already had before. Also, it's already part of sdk-nrf in its current form (and it works "well enough" to unblock certain work), so we do not have the same pressure as before to discuss technical aspects as before :-)

nika-nordic replied:
What I am afraid of is that scheme of clearing events in nrfx drivers and nrfy layer during initialization is established for quite some time. Rationale is that during initialization driver applies clear & known hardware configuration, to avoid situation in which leftover peripheral register contents alter the driver behavior.

If we decide to change the nrfx behavior in this manner (that is to handle events that might be triggered before driver initialization and shall be considered as valid) IMO such change should be done systematically across all drivers, instead of doing it ad-hoc here

hubertmis replied:
IPC signaling peripherals might be a special case requiring a different approach. But let's discuss it in the "follow up PR" scope mentioned by @gmarull

hubertmis commented on drivers/mbox/mbox_nrf_bellboard_local.c:
```diff
@@ -0,0 +1,169 @@
+			}
+
+			*evt_enabled_mask &= ~BIT(id);
+			nrf_bellboard_int_disable(bellboard, i, BIT(id));
```
Similarly to VEVIF, I don't think interrupts should get disabled unconditionally, because other still active channels may be assigned to the same interrupt.

gmarull replied:
Note that here we’re never disabling IRQ, but changing the interrupt enabled masks.

gmarull commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:
```diff
@@ -8,8 +8,11 @@
+#include "nrf54h20pdk_nrf54h20-ipc_conf.dtsi"
 #include "nrf54h20pdk_nrf54h20-pinctrl.dtsi"
 
+/delete-node/ &cpusec_cpurad_ipc;
```
I think deleting is preferred, because this actually doesn’t exist for this core. This could be avoided by not creating any “common” file I guess.

gmarull commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:
```diff
@@ -15,7 +105,7 @@
-		ram3x_dma_region: memory@2fc07000 {
+		shared_ram3x_region: memory@2fc07000 {
 			compatible = "nordic,owned-memory";
 			reg = <0x2fc07000 DT_SIZE_K(4)>;
```
@57300

57300 replied:
This is not Final Product, so the 4 KiB boundary applies. @gmarull Up to you if you want to add the `TODO`.

gmarull commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:
```diff
@@ -6,6 +6,96 @@
+
+		shared_ram20_region: memory@2f88f000 {
+			compatible = "nordic,owned-memory";
+			reg = <0x2f88f000 DT_SIZE_K(4)>;
```
Note this is all “soc1” or EngA

57300 commented on dts/common/nordic/nrf54h20_enga.dtsi:
```diff
@@ -281,6 +306,15 @@
+					nordic,tasks-mask = <0xfffffff0>;
+				};
+
 				cpuppr_clic: interrupt-controller@1000 {
```
This discussion can continue in #69281

nordicjm commented on samples/drivers/mbox/remote/CMakeLists.txt:
```diff
@@ -15,7 +15,8 @@ if(("${BOARD}" STREQUAL "nrf5340dk_nrf5340_cpunet") OR
   ("${BOARD}" STREQUAL "lpcxpresso55s69_cpu1") OR
-  ("${BOARD}" STREQUAL "adp_xc7k_ae350"))
+  ("${BOARD}" STREQUAL "adp_xc7k_ae350") OR
+  ("${BOARD}" STREQUAL "nrf54h20pdk_nrf54h20_cpuppr"))
```
Needs converting to hwmv2

gmarull replied:
done

gmarull committed:
samples: drivers: mbox: fix hwmv2 issues

Some board names were not ported to HWMv2 scheme.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
@nordicjm @anangl @57300 ping
